### PR TITLE
Adding `T: Bake` bound on `ZeroVec`'s `Bake`

### PR DESCRIPTION
**Pull request by robertbastian**
Based on #3450 

This bound is currently unused, but will be useful as part of #1935

### CODE REVIEW

File: utils/zerovec/src/ule/unvalidated.rs [hunk 1]
```diff
@@ -412,6 +419,28 @@ impl<'de> serde::Deserialize<'de> for UnvalidatedChar {
...
+            Ok(ch) => {
+                env.insert("zerovec");
+                let ch = ch.bake(env);
+                databake::quote! {
+                    zerovec::ule::UnvalidatedChar::from_char(#ch)
+                }
+            }
+            Err(_) => {
+                env.insert("zerovec");
+                let u24 = u32::from_le_bytes([self.0[0], self.0[1], self.0[2], 0]);
```
Comment (skius) [inline utils/zerovec/src/ule/unvalidated.rs:435]: Is there no `bake` here intentionally?

Comment (robertbastian) [reply utils/zerovec/src/ule/unvalidated.rs:435]: Somewhat, it's more code for the same result.

**Comment by Manishearth** [comment]
Not opposed. Bit worried about how we can manage this for VZV, if at all, but we don't need to.

This is a breaking change in databake, though. I think it would be nice if we could mitigate that breakage for one release (let ICU4X's databake dep be a range dep for 1.3). Not blocking, and we can see what it looks like when we do a release.

**Comment by robertbastian** [comment]
For VZV we'd probably want a `&T: Bake` bound or something. That's further out though.

The breaking in databake is in the other PR, no?

**Comment by Manishearth** [comment]
Oh, no, it's breaking `zerovec`, nvm

**Comment by robertbastian** [comment]
Meh, nobody uses the `databake` feature (at least there are no `databake` users on crates.io other than us).

**Review by sffc — APPROVED** [review]

File: utils/tinystr/src/databake.rs [hunk 2]
```diff
@@ -15,7 +16,33 @@ impl<const N: usize> Bake for TinyAsciiStr<N> {
     }
 }
 
+impl<const N: usize> databake::Bake for UnvalidatedTinyAsciiStr<N> {
```
Comment (sffc) [inline utils/tinystr/src/databake.rs:19]: Nit: Please mention these impls in the PR title.

File: utils/zerovec/src/zerovec/databake.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ use databake::*;
 
 impl<T> Bake for ZeroVec<'_, T>
 where
-    T: AsULE + ?Sized,
+    T: AsULE + ?Sized + Bake,
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/zerovec/src/zerovec/databake.rs:11]: Yes, though we need to figure out if we want to follow semver and release a breaking zerovec bump.

**Pull request merged**